batchflow/plot.py

```diff
@@ -0,0 +1,830 @@
+    """ Plotting backend for matplotlib.
+
+    Overall idea
+    ------------
+    Simply provide data, plot mode and parameters to the `plot` initialization
+    and the class takes care of redirecting params to methods they are meant for.
+
+    The logic behind the process is the following:
+    1. Parse data:
+        - Calculate subplots shapes data.
+        - Put provided arrays into double nested list.
+          Nestedness levels define subplot and layer data order correspondingly.
+        - Infer images combination mode.
+        - Calculate indices corrections for empty subplots.
+    2. Parse figure axes if provided, else create them with either parsed parameters or inferred ones.
+    3. Obtain default params for chosen mode and merge them with provided params.
+    4. For every axis-data pair:
+        - If no data provided for axis, set if off.
+        - Else filter params relevant for ax, plot data relevant to the ax and annotate it.
+    6. Save figure if needed.
+
+    Data display scenarios
+    ----------------------
+    1. The simplest one if when one provide a single data array.
+    2. A more advanced use case is when one provide a list of arrays:
+       a. Images are put on same axis and overlaid: data=[image_1, image_2] (combine='overlay' is default);
+       b. Images are put on separate axes: data=[image_1, image_2], combine='separate';
+    3. The most complex scenario is when one wish to display images in a 'mixed' manner.
+       For example, to overlay first two images but to display the third one separately, one must use:
+       data = [[image_1, image_2], image_3] (combine='mixed' is set automatically if data is double-nested);
+
+    The order of arrrays inside the double-nested structure basically declares, which of them belong to the same axis
+    and therefore should be rendered one over another, and which must be displayed separately.
+
+    Note that in general parameters should resemble data nestedness level.
+    That allows binding axes and parameters that correspond each other.
+    However, it's possible for parameter to be a single item — in that case it's shared across all subplots and layers.
+
+    Advanced parameters managing
+    ----------------------------
+    The list of parameters expected by specific plot method is rather short.
+    But there is a way to provide parameter to a plot method, even if it's not hard-coded.
+    One must use specific prefix for that.
+    Address docs of `imshow`, `wiggle`, `hist`, `curve` and `annotate_axis` for details.
+
+    This also allows one to pass arguments of the same name for different plotting steps.
+    E.g. `plt.set_title` and `plt.set_xlabel` both require `fontsize` argument.
+    Providing `{'fontsize': 30}` in kwargs will affect both title and x-axis labels.
+    To change parameter for title only, one can provide {'title_fontsize': 30}` instead.
+    """
+    def __init__(self, data=None, combine='overlay', mode='imshow', **kwargs):
```

Comment:
This docstring is unavailable from the outside: move the `__init__` explanation and parameters to the class docstring